Unable to use basic filter with collection (was: Inconsistent behavior with SQL connector and message split)

## This is a...

<pre><code>
[ ] Feature request
[ ] Regression (a behavior that used to work and stopped working in a new release)
[X] Bug report  <!-- Please search GitHub for a similar issue or PR before submitting -->
[ ] Documentation issue or request
</code></pre>

## Description

I have following integration:

```
1. SQL - Periodic invocation - Select * from contact
2. AMQ - Publish messages to some queue
```

In this scenario, the list from SQL connector is automatically split into multiple messages for some reason even when I didn't specify the `split` step. So when I have 2 contacts in the database, I receive 2 messages in the queue.

Now I want to add a basic filter to the integration - since the split magic is being done, I might not need the explicit "split" step, so I create:

```
1. SQL - Periodic invocation - Select * from contact
2. Basic filter - last_name contains something (now I know something is wrong because I don't have the autocomplete for the field)
3. AMQ - Publish messages to some queue
```  

Obviously, the integration fails with:

```
org.apache.camel.spring.boot.CamelSpringBootInitializationException: java.lang.IllegalArgumentException: java.lang.IllegalStateException: No step properties defined for rule filter step
	at org.apache.camel.spring.boot.RoutesCollector.onApplicationEvent(RoutesCollector.java:250) ~[camel-spring-boot-2.21.0.fuse-730042.jar!/:2.21.0.fuse-730042]
	at org.apache.camel.spring.boot.RoutesCollector.onApplicationEvent(RoutesCollector.java:57) ~[camel-spring-boot-2.21.0.fuse-730042.jar!/:2.21.0.fuse-730042]
	at org.springframework.context.event.SimpleApplicationEventMulticaster.doInvokeListener(SimpleApplicationEventMulticaster.java:172) ~[spring-context-4.3.19.RELEASE.jar!/:4.3.19.RELEASE]
	at org.springframework.context.event.SimpleApplicationEventMulticaster.invokeListener(SimpleApplicationEventMulticaster.java:165) ~[spring-context-4.3.19.RELEASE.jar!/:4.3.19.RELEASE]
	at org.springframework.context.event.SimpleApplicationEventMulticaster.multicastEvent(SimpleApplicationEventMulticaster.java:139) ~[spring-context-4.3.19.RELEASE.jar!/:4.3.19.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.publishEvent(AbstractApplicationContext.java:393) ~[spring-context-4.3.19.RELEASE.jar!/:4.3.19.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.publishEvent(AbstractApplicationContext.java:347) ~[spring-context-4.3.19.RELEASE.jar!/:4.3.19.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.finishRefresh(AbstractApplicationContext.java:883) ~[spring-context-4.3.19.RELEASE.jar!/:4.3.19.RELEASE]
	at org.springframework.boot.context.embedded.EmbeddedWebApplicationContext.finishRefresh(EmbeddedWebApplicationContext.java:144) ~[spring-boot-1.5.16.RELEASE.jar!/:1.5.16.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:546) ~[spring-context-4.3.19.RELEASE.jar!/:4.3.19.RELEASE]
	at org.springframework.boot.context.embedded.EmbeddedWebApplicationContext.refresh(EmbeddedWebApplicationContext.java:122) ~[spring-boot-1.5.16.RELEASE.jar!/:1.5.16.RELEASE]
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:693) [spring-boot-1.5.16.RELEASE.jar!/:1.5.16.RELEASE]
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:360) [spring-boot-1.5.16.RELEASE.jar!/:1.5.16.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:303) [spring-boot-1.5.16.RELEASE.jar!/:1.5.16.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1118) [spring-boot-1.5.16.RELEASE.jar!/:1.5.16.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1107) [spring-boot-1.5.16.RELEASE.jar!/:1.5.16.RELEASE]
	at io.syndesis.example.Application.main(Application.java:13) [classes!/:na]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_191]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_191]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_191]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_191]
	at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:48) [project-0.1-SNAPSHOT.jar:na]
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:87) [project-0.1-SNAPSHOT.jar:na]
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:50) [project-0.1-SNAPSHOT.jar:na]
	at org.springframework.boot.loader.PropertiesLauncher.main(PropertiesLauncher.java:595) [project-0.1-SNAPSHOT.jar:na]
Caused by: java.lang.IllegalArgumentException: java.lang.IllegalStateException: No step properties defined for rule filter step
	at io.syndesis.integration.runtime.sb.IntegrationRuntimeAutoConfiguration$1.beforeApplicationStart(IntegrationRuntimeAutoConfiguration.java:94) ~[integration-runtime-springboot-1.6-SNAPSHOT.jar!/:1.6-SNAPSHOT]
	at org.apache.camel.spring.boot.RoutesCollector.onApplicationEvent(RoutesCollector.java:152) ~[camel-spring-boot-2.21.0.fuse-730042.jar!/:2.21.0.fuse-730042]
	... 24 common frames omitted
Caused by: java.lang.IllegalStateException: No step properties defined for rule filter step
	at io.syndesis.integration.runtime.handlers.RuleFilterStepHandler.getFilterExpression(RuleFilterStepHandler.java:51) ~[integration-runtime-1.6-SNAPSHOT.jar!/:1.6-SNAPSHOT]
	at io.syndesis.integration.runtime.handlers.AbstractFilterStepHandler.handle(AbstractFilterStepHandler.java:36) ~[integration-runtime-1.6-SNAPSHOT.jar!/:1.6-SNAPSHOT]
	at io.syndesis.integration.runtime.handlers.RuleFilterStepHandler.handle(RuleFilterStepHandler.java:30) ~[integration-runtime-1.6-SNAPSHOT.jar!/:1.6-SNAPSHOT]
	at io.syndesis.integration.runtime.IntegrationRouteBuilder.configureFlow(IntegrationRouteBuilder.java:174) ~[integration-runtime-1.6-SNAPSHOT.jar!/:1.6-SNAPSHOT]
	at io.syndesis.integration.runtime.IntegrationRouteBuilder.configure(IntegrationRouteBuilder.java:104) ~[integration-runtime-1.6-SNAPSHOT.jar!/:1.6-SNAPSHOT]
	at org.apache.camel.builder.RouteBuilder.checkInitialized(RouteBuilder.java:462) ~[camel-core-2.21.0.fuse-730042.jar!/:2.21.0.fuse-730042]
	at org.apache.camel.builder.RouteBuilder.configureRoutes(RouteBuilder.java:402) ~[camel-core-2.21.0.fuse-730042.jar!/:2.21.0.fuse-730042]
	at org.apache.camel.builder.RouteBuilder.addRoutesToCamelContext(RouteBuilder.java:383) ~[camel-core-2.21.0.fuse-730042.jar!/:2.21.0.fuse-730042]
	at org.apache.camel.impl.DefaultCamelContext$1.call(DefaultCamelContext.java:1027) ~[camel-core-2.21.0.fuse-730042.jar!/:2.21.0.fuse-730042]
	at org.apache.camel.impl.DefaultCamelContext$1.call(DefaultCamelContext.java:1024) ~[camel-core-2.21.0.fuse-730042.jar!/:2.21.0.fuse-730042]
	at org.apache.camel.impl.DefaultCamelContext.doWithDefinedClassLoader(DefaultCamelContext.java:3270) ~[camel-core-2.21.0.fuse-730042.jar!/:2.21.0.fuse-730042]
	at org.apache.camel.impl.DefaultCamelContext.addRoutes(DefaultCamelContext.java:1024) ~[camel-core-2.21.0.fuse-730042.jar!/:2.21.0.fuse-730042]
	at io.syndesis.integration.runtime.sb.IntegrationRuntimeAutoConfiguration$1.beforeApplicationStart(IntegrationRuntimeAutoConfiguration.java:92) ~[integration-runtime-springboot-1.6-SNAPSHOT.jar!/:1.6-SNAPSHOT]
	... 25 common frames omitted

```

Probably the culprit is https://github.com/syndesisio/syndesis/issues/562 , but I think it does not make sense now since we have the dedicated `split`

--- a/app/ui/src/app/integration/edit-page/flow-functions.ts
+++ b/app/ui/src/app/integration/edit-page/flow-functions.ts
@@ -42,10 +42,19 @@ function setStepId(step: Step) {
 
 /**
  * Validate and add/correct items in a flow's step array
+ *  * Sets step IDs
+ *  * Strips out invalid/unconfigured steps
  * @param flow
  */
 function validateFlowSteps(flow: Flow) {
-  return { ...flow, ...{ steps: (flow.steps || []).map(setStepId) } };
+  return {
+    ...flow,
+    ...{
+      steps: (flow.steps || [])
+        .map(setStepId)
+        .filter(s => typeof s.stepKind !== 'undefined'),
+    },
+  };
 }
 
 /**

--- a/app/ui/src/app/store/step/step.store.ts
+++ b/app/ui/src/app/store/step/step.store.ts
@@ -40,23 +40,26 @@ export class StepStore {
         description: 'Map fields from the input type to the output type.',
         stepKind: DATA_MAPPER,
         properties: {},
-        configuredProperties: undefined
+        configuredProperties: undefined,
       },
       true,
       false
     ),
-    StepStore.requiresOutputDataShape({
-      id: undefined,
-      connection: undefined,
-      action: undefined,
-      name: 'Basic Filter',
-      description:
-        'Continue the integration only if criteria you specify in simple input fields are met. Suitable for' +
-        ' most integrations.',
-      stepKind: BASIC_FILTER,
-      properties: undefined,
-      configuredProperties: undefined
-    }),
+    StepStore.requiresOutputDataShape(
+      {
+        id: undefined,
+        connection: undefined,
+        action: undefined,
+        name: 'Basic Filter',
+        description:
+          'Continue the integration only if criteria you specify in simple input fields are met. Suitable for' +
+          ' most integrations.',
+        stepKind: BASIC_FILTER,
+        properties: undefined,
+        configuredProperties: undefined,
+      },
+      true
+    ),
     StepStore.requiresOutputDataShape({
       id: undefined,
       connection: undefined,
@@ -66,9 +69,9 @@ export class StepStore {
       description:
         'Upload or create a Freemarker, Mustache or Velocity template to define consistent output data.',
       configuredProperties: undefined,
-      properties: undefined
+      properties: undefined,
     }),
-    {
+    StepStore.noCollectionSupport({
       id: undefined,
       connection: undefined,
       action: undefined,
@@ -86,11 +89,11 @@ $\{in.header.type\} == 'widget' // Evaluates true when type = widget
 $\{in.body.title\} // Evaluates true when body contains title.
 `,
           required: true,
-          rows: 10
-        }
+          rows: 10,
+        },
       },
-      configuredProperties: undefined
-    },
+      configuredProperties: undefined,
+    }),
     {
       id: undefined,
       connection: undefined,
@@ -103,18 +106,18 @@ $\{in.body.title\} // Evaluates true when body contains title.
         contextLoggingEnabled: {
           type: 'boolean',
           displayName: 'Message Context',
-          required: false
+          required: false,
         },
         bodyLoggingEnabled: {
           type: 'boolean',
           displayName: 'Message Body',
-          required: false
+          required: false,
         },
         customText: {
           type: 'string',
           displayName: 'Custom Text',
-          required: false
-        }
+          required: false,
+        },
         /*
         loggingLevel: {
           type: 'select',
@@ -129,7 +132,7 @@ $\{in.body.title\} // Evaluates true when body contains title.
             {value: 'TRACE', label: 'TRACE'}],
         },
         */
-      }
+      },
     },
     StepStore.requiresOutputDataShape({
       id: undefined,
@@ -139,7 +142,7 @@ $\{in.body.title\} // Evaluates true when body contains title.
       description: 'Process each item in a set of data individually',
       stepKind: SPLIT,
       properties: {},
-      configuredProperties: undefined
+      configuredProperties: undefined,
     }),
     StepStore.requiresConsistentSplitAggregate({
       id: undefined,
@@ -149,8 +152,8 @@ $\{in.body.title\} // Evaluates true when body contains title.
       description: 'End processing items in a foreach',
       stepKind: AGGREGATE,
       properties: {},
-      configuredProperties: undefined
-    })
+      configuredProperties: undefined,
+    }),
     /*
     {
       id: undefined,
@@ -255,7 +258,7 @@ $\{in.body.title\} // Evaluates true when body contains title.
               properties: properties,
               extension: extension,
               action: action,
-              configuredProperties: undefined
+              configuredProperties: undefined,
             });
           }
         }
@@ -327,13 +330,13 @@ $\{in.body.title\} // Evaluates true when body contains title.
           []
             .concat(previousSteps)
             .reverse()
-            .find(s => StepStore.dataShapeExists(s))
+            .find(s => StepStore.dataShapeExists(s)),
         ];
       }
       if (!anySubsequent) {
         // only test the next subsequent step that has a data shape
         subsequentSteps = [
-          subsequentSteps.find(s => StepStore.dataShapeExists(s, true))
+          subsequentSteps.find(s => StepStore.dataShapeExists(s, true)),
         ];
       }
       return (
@@ -362,14 +365,57 @@ $\{in.body.title\} // Evaluates true when body contains title.
     }
   }
 
-  static requiresOutputDataShape(obj: StepKind): StepKind {
+  static hasPrecedingCollection(previousSteps: Step[]) {
+    const previousDataShape = []
+      .concat(previousSteps)
+      .reverse()
+      .find(s => StepStore.dataShapeExists(s));
+    return (
+      previousDataShape &&
+      previousDataShape.action &&
+      previousDataShape.action.description &&
+      previousDataShape.action.descriptor.outputDataShape &&
+      previousDataShape.action.descriptor.outputDataShape.metadata &&
+      previousDataShape.action.descriptor.outputDataShape.metadata.variant ===
+        'collection'
+    );
+  }
+
+  static noCollectionSupport(obj: StepKind) {
     obj.visible = (
       position: number,
       previousSteps: Array<Step>,
       subsequentSteps: Array<Step>
     ) => {
-      return StepStore.stepsHaveOutputDataShape(previousSteps);
+      return !StepStore.hasPrecedingCollection(previousSteps);
     };
+    return obj;
+  }
+
+  static requiresOutputDataShape(
+    obj: StepKind,
+    noCollectionSupport = false
+  ): StepKind {
+    if (noCollectionSupport) {
+      obj.visible = (
+        position: number,
+        previousSteps: Array<Step>,
+        subsequentSteps: Array<Step>
+      ) => {
+        return (
+          StepStore.stepsHaveOutputDataShape(previousSteps) &&
+          !StepStore.hasPrecedingCollection(previousSteps)
+        );
+      };
+    } else {
+      obj.visible = (
+        position: number,
+        previousSteps: Array<Step>,
+        subsequentSteps: Array<Step>
+      ) => {
+        return StepStore.stepsHaveOutputDataShape(previousSteps);
+      };
+    }
     return obj;
   }
 
@@ -379,21 +425,31 @@ $\{in.body.title\} // Evaluates true when body contains title.
       previousSteps: Array<Step>,
       subsequentSteps: Array<Step>
     ) => {
-      const countPreviousSplit = previousSteps.filter(s => s.stepKind == SPLIT).length;
-      const countPreviousAggregate = previousSteps.filter(s => s.stepKind == AGGREGATE).length;
+      const countPreviousSplit = previousSteps.filter(s => s.stepKind == SPLIT)
+        .length;
+      const countPreviousAggregate = previousSteps.filter(
+        s => s.stepKind == AGGREGATE
+      ).length;
 
       if (countPreviousSplit <= countPreviousAggregate) {
         return false;
       }
 
-      const positionNextSplit = subsequentSteps.findIndex(s => s.stepKind == SPLIT);
-      const positionNextAggregate = subsequentSteps.findIndex(s => s.stepKind == AGGREGATE);
+      const positionNextSplit = subsequentSteps.findIndex(
+        s => s.stepKind == SPLIT
+      );
+      const positionNextAggregate = subsequentSteps.findIndex(
+        s => s.stepKind == AGGREGATE
+      );
 
       if (positionNextSplit === -1) {
         return positionNextAggregate === -1;
       }
 
-      return positionNextAggregate === -1 || positionNextSplit < positionNextAggregate;
+      return (
+        positionNextAggregate === -1 ||
+        positionNextSplit < positionNextAggregate
+      );
     };
     return obj;
   }

